[FIX] mail: convert cloud-stored attachments to links in outgoing emails

nd-dew opened the pull request:
Before this commit, when sending emails with attachments stored in the cloud, the attachments's links were not included in the email body, as we only included the links for attachments exceeding the max email size.

With this commit, we ensure that all attachments stored in the cloud are converted to links in the email body and included in the the email.

opw-4717083

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/208424.png)](https://mergebot.odoo.com/odoo/odoo/pull/208424)

HydrionBurst commented on addons/mail/models/mail_mail.py:
```diff
@@ -497,6 +497,14 @@ def _prepare_outgoing_list(self, mail_server=False, recipients_follower_status=N
                                                                 {'attachments': record_owned_attachments})
                 body = tools.mail.append_content_to_html(body, attachments_links, plaintext=False)
                 attachments -= record_owned_attachments
+        cloud_attachments = attachments.sudo().filtered(
+            lambda a: a.type == 'cloud_storage' and a.res_model and a.res_id and a.res_model != 'mail.message'
+        )
+        if cloud_attachments:
```
Thanks for locating the problem. It seems I shouldn't directly remove my code when solving the [conflicts](https://github.com/odoo/odoo/pull/157249#issuecomment-2099873109)

I propose we don't use `'cloud_storage'` here, since the module `cloud_storage` is not a commonly installed module.

Also it would be cleaner to groupby `record_owned_attachments` into 2 categories
1. url_attachments: `a.url and a.url.startswith(('https://', 'http://', 'ftp://'))` => always use the `append_content_to_html`
2. binary_attachments: => old logic

So an attachment will only be added to the email once.

nd-dew replied:
Appreciate the feedback — and totally agree this should be scoped to the `cloud_storage` module. I’m in the middle of moving it there, just didn’t get to clean it up yet — still draft 🙃

As for refactoring the grouping logic: yeah that would be cleaner, but feels a bit outside the scope of this fix (in bugfix lane, so trying to keep changes minimal). Definitely worth considering for a future cleanup though!

nd-dew commented:
Since this module is not heavily used I moved the fix from mail_mail to cloud_storage module

HydrionBurst commented on addons/cloud_storage/models/mail_mail.py:
```diff
@@ -0,0 +1,49 @@
+        for mail, outgoing in zip(self, outgoing_list):
+            cloud_attachments = mail.attachment_ids.filtered(
+                lambda attachment: attachment.type == 'cloud_storage'
+                    and attachment.res_model != 'mail.message'
```
I think this condition is not useful.
Technically, now cloud stroage attachments won't be attached to `mail.message`. But even if we can attachment them. We should change the attachment to the `'mail.mail_attachment_links'` becasue otherwise the `raw` of these attachments will still be empty.

HydrionBurst commented on addons/cloud_storage/models/mail_mail.py:
```diff
@@ -0,0 +1,49 @@
+            link_ids = {int(link) for link in re.findall(r'/web/(?:content|image)/([0-9]+)', outgoing['body'])}
+            if link_ids:
+                cloud_attachments -= self.env['ir.attachment'].browse(list(link_ids))
+
```
```py
if not cloud_attachments:
    continue
```

HydrionBurst commented on addons/cloud_storage/models/mail_mail.py:
```diff
@@ -0,0 +1,49 @@
+        )
+
+        # Turn remaining attachments into links appended to an email if they are cloud attachments
+        for mail, outgoing in zip(self, outgoing_list):
```
This logic is not correct
you can find `self.ensure_one()` in the `super()._prepare_outgoing_list(`

`outgoing_list` are emails to different people for the same mail record. This line only handles the first `outgoing`.

BTW. The `outgoing['attachments']` are shared list, you can replace its data it in place for the first one.` But `outgoing['body']` are personalized.

HydrionBurst commented on addons/cloud_storage_google/tests/test_cloud_storage_google_attachment_controller.py:
```diff
@@ -68,3 +69,34 @@ def test_cloud_storage_google_attachment_upload(self):
+        self.env["ir.config_parameter"].set_param("cloud_storage_min_file_size", 1)
+
+        partner = self.env["res.partner"].create({"name": "Cloud Test Partner", "email": "cloud@test.com"})
+        cloud_attachment = self.env["ir.attachment"].create({
```
Maybe also create another test for attachments
1. a large non cloud storage attachment
2. cloud stroage attachment

Check if the both attachment links are added and without duplicate.

HydrionBurst commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -1656,7 +1656,7 @@ def test_message_post(self):
         # enable_logging = self.cr._enable_logging() if self.warm else nullcontext()
         # with self.assertQueryCount(employee=63), enable_logging:
-        with self.assertQueryCount(employee=63):
+        with self.assertQueryCount(employee=64):
```
Could you explain a little why the query count is increased?

HydrionBurst commented:
> Since this module is not heavily used I moved the fix from mail_mail to cloud_storage module

I think the current `_prepare_outgoing_list` in `mail`
Pure url attachments whose `attachment.raw is False` may also not be attached in the email. It would make sense if we improve that from the `mail`. So that all `not attachment.file_size and attachment.url` attachments will be attached to as clip instead of binary.

Maybe @pyduf can give some suggestions.

pyduf commented:
> > Since this module is not heavily used I moved the fix from mail_mail to cloud_storage module
> 
> I think the current `_prepare_outgoing_list` in `mail` Pure url attachments whose `attachment.raw is False` may also not be attached in the email. It would make sense if we improve that from the `mail`. So that all `not attachment.file_size and attachment.url` attachments will always be attached as clips instead of binary files.
> 
> Maybe @pyduf can give some suggestions.

Hello @HydrionBurst and @nd-dew ,

Yes I think it may be better to handle this/define a hook in mail for converting attachment into links to centralize that and avoid handling it more than necessary (handling it before the split into recipients allow to handle once instead of multiple times). 

Without a lot of thinking, we could define a method like _prepare_attachments that would determine which attachment is to be converted as link. The base code would be the code about converting big attachments as links and in the cloud_storage module, that method would be overridden to convert “cloud_storage” attachment type into links as well.

It’s a long time so I may missing something.  @tde-banana-odoo what do you think ? Thanks.

HydrionBurst commented:
I still think we can simply append all url attachments `a.url and not a.file_size` as clips with odoo url (/web/content...). Because
1. we rarely use url attachments 🤔, especially sent them in emails.
2. even if we have simple url attachments which are totally ignored in the current code when sending emails, the overhead is one request per download, and odoo is only used for redirection.  In the use case of email, considering the download frequency, it should be quite acceptable. And administrators can even redirects customers to a new url by simply changing the `attachment.url`
3. it is very simple without any code customized in `cloud_storage` module since a cloud storage attachment must be a url attachment.

nd-dew commented:
Thanks for the input 😊

> I still think we can simply append all url attachments a.url and url.startswith(('https://', 'http://', 'ftp://') and not a.file_size as clips with odoo url (/web/content...). 

I think @HydrionBurst’s suggestion makes a lot of sense here and would let us solve this bug in a much simpler and more generic way. 

Let me put that roughly in code. to quickly see if that's more/less what you had in mind

nd-dew commented on addons/mail/models/mail_mail.py:
```diff
@@ -490,6 +490,15 @@ def _prepare_outgoing_list(self, mail_server=False, recipients_follower_status=N
             max_email_size_bytes = (mail_server or self.env['ir.mail_server']
                                     ).sudo()._get_max_email_size() * 1024 * 1024
+            # Add links for all URL attachments without file_size
+            url_attachments = attachments.sudo().filtered(
+                lambda a: a.url and not a.file_size and a.url.startswith(('http://', 'https://', 'ftp://'))
+            )
```
not sure if that's grabbing all it should

HydrionBurst commented:
Let me know if there's anything I can do to help move this PR forward 😃

nd-dew commented:
Dived a bit more: coping happens only in the `mass_mail` `composition_mode`. Perhaps I should switch to the comment mode in my test to better fit the test flow.

nd-dew commented:
Updated, tested, cleaned and squashed. 
Test sanity check ok: pass with fix, fail without

Note about the complex test case I was designing before:
Sending 2 attachments, small and large doesn't make sense, because if they are both in the same message the current logic will just detect that total size is exceeding the limit and convert them all into links
